src/app/users/login.ts

```diff
@@ -0,0 +1,130 @@
+import {
+  getFailedLoginAttemptPerIpLimits,
+  getFailedLoginAttemptPerPhoneLimits,
+} from "@config/app"
+import { CouldNotFindUserFromPhoneError } from "@domain/errors"
+import { RateLimitPrefix } from "@domain/rate-limit"
+import {
+  RateLimiterExceededError,
+  UserLoginIpRateLimiterExceededError,
+  UserLoginPhoneRateLimiterExceededError,
+} from "@domain/rate-limit/errors"
+import { createToken } from "@services/jwt"
+import { UsersRepository } from "@services/mongoose"
+import { PhoneCodesRepository } from "@services/mongoose/phone-code"
+import { getCarrier } from "@services/phone-provider"
+import { RedisRateLimitService } from "@services/rate-limit"
+import { isTestAccount } from "."
+
+export const login = async ({
+  phone,
+  code,
+  logger,
+  ip,
+}: {
+  phone: PhoneNumber
+  code: PhoneCode
+  logger: Logger
+  ip: IpAddress
+}): Promise<JWTToken | ApplicationError> => {
+  const subLogger = logger.child({ topic: "login" })
+
+  {
+    const limitOk = await checkFailedLoginAttemptPerIpLimits(ip)
+    if (limitOk instanceof Error) return limitOk
+  }
+
+  {
+    const limitOk = await checkFailedLoginAttemptPerPhoneLimits(phone)
+    if (limitOk instanceof Error) return limitOk
+  }
+
+  // TODO:
+  // add fibonachi on failed login
+  // https://github.com/animir/node-rate-limiter-flexible/wiki/Overall-example#dynamic-block-duration
+
+  const validCode = await isCodeValid({ phone, code })
+  if (validCode instanceof Error) return validCode
+
+  // reseting the limiter for this phone
+  // limiterLoginAttempt.delete(phone) // no need to await the promise
```

Comment:
if it is not necessary please remove the comments

Reply:
it will put a limit on how frequently someone can log in. previously it was possible to login infinitely on a single phone number. but this reset is actually harmful after thinking more about it. so I'll delete.